### ISSUE TITLE
Release OpenTSLM models and datasets on Hugging Face

### ISSUE THREAD

**Issue by NielsRogge** [StanfordBDHG/OpenTSLM#28]
Hi @StanfordBDHG 🤗

Niels here from the open-source team at Hugging Face. I discovered your work through Hugging Face's daily papers as yours got featured: https://huggingface.co/papers/2510.02410.
The paper page lets people discuss about your paper and lets them find artifacts about it (your models, datasets or demo for instance), you can also claim
the paper as yours which will show up on your public profile at HF, add Github and project page URLs.

It'd be great to make the OpenTSLM model checkpoints (OpenTSLM-SoftPrompt, OpenTSLM-Flamingo) and the new datasets (HAR-CoT, Sleep-CoT, ECG-QA-CoT) available on the 🤗 hub, to improve their discoverability/visibility.
We can add tags so that people find them when filtering https://huggingface.co/models and https://huggingface.co/datasets.

## Uploading models

See here for a guide: https://huggingface.co/docs/hub/models-uploading.

In this case, we could leverage the [PyTorchModelHubMixin](https://huggingface.co/docs/huggingface_hub/package_reference/mixins#huggingface_hub.PyTorchModelHubMixin) class which adds `from_pretrained` and `push_to_hub` to any custom `nn.Module`. Alternatively, one can leverages the [hf_hub_download](https://huggingface.co/docs/huggingface_hub/en/guides/download#download-a-single-file) one-liner to download a checkpoint from the hub.

We encourage researchers to push each model checkpoint to a separate model repository, so that things like download stats also work. We can then also link the checkpoints to the paper page.

## Uploading dataset

Would be awesome to make the datasets available on 🤗 , so that people can do:

```python
from datasets import load_dataset

dataset = load_dataset("your-hf-org-or-username/your-dataset")
```
See here for a guide: https://huggingface.co/docs/datasets/loading.

Besides that, there's the [dataset viewer](https://huggingface.co/docs/hub/en/datasets-viewer) which allows people to quickly explore the first few rows of the data in the browser.

Let me know if you're interested/need any help regarding this!

Cheers,

Niels
ML Engineer @ HF 🤗

**Comment by code-rishav** [comment]
```
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM
from peft import get_peft_model, LoraConfig

path = "/content/softprompt-gemma_3_270m-har.pt"
ckpt = torch.load(path, map_location="cpu", weights_only=False)

# Extract LoRA parts
lora_state = ckpt["lora_state"] if "lora_state" in ckpt else ckpt
lora_cfg = ckpt.get("lora_config", None)
if isinstance(lora_cfg, dict):
    print(lora_cfg)
    lora_cfg = lora_cfg['default']

# If lora_config is missing, infer r/targets from keys
if lora_cfg is None:
    r = None
    targets = set()
    for k, v in lora_state.items():
        if "lora_A" in k and hasattr(v, "shape") and v.ndim == 2:
            r = v.shape[0]
        for name in ["q_proj","k_proj","v_proj","o_proj","gate_proj","up_proj","down_proj"]:
            if f".{name}." in k:
                targets.add(name)
    if r is None:
        r = 8
    if not targets:
        targets = {"q_proj","k_proj","v_proj","o_proj"}
    lora_cfg = LoraConfig(
        r=r, lora_alpha=2*r, lora_dropout=0.05,
        target_modules=sorted(list(targets)),
        task_type="CAUSAL_LM"
    )

# Load base model (use -it for instruction prompts)
device = "cuda" if torch.cuda.is_available() else "cpu"
dtype = torch.bfloat16 if torch.cuda.is_available() else torch.float32
base_id = "google/gemma-3-270m-it"  # or "google/gemma-3-270m"
tok = AutoTokenizer.from_pretrained(base_id)
base = AutoModelForCausalLM.from_pretrained(base_id, torch_dtype=dtype).to(device)

# Wrap with LoRA and load adapter weights
peft_model = get_peft_model(base, lora_cfg).to(device)
missing, unexpected = peft_model.load_state_dict(lora_state, strict=False)
peft_model.eval()

# Chat-style prompting (uses tokenizer chat template if available)
messages = [
    {"role": "system", "content": "You are a helpful assistant for analytics on user events."},
    {"role": "user", "content": "Summarize search and cart order patterns from this log: ..."}
]

# Build inputs as a mapping with attention_mask
inputs = tok.apply_chat_template(
    messages,
    tokenize=True,
    add_generation_prompt=True,
    return_tensors="pt",
    return_dict=True,           # ensures {"input_ids","attention_mask"}
)
inputs = {k: v.to(peft_model.device) for k, v in inputs.items()}

with torch.no_grad():
    gen_ids = peft_model.generate(
        **inputs,
        max_new_tokens=256,
        do_sample=True, temperature=0.7, top_p=0.9,
        pad_token_id=tok.eos_token_id,   # or tok.pad_token_id if defined
        eos_token_id=tok.eos_token_id,
    )

prompt_len = inputs["input_ids"].shape[-1]
text = tok.batch_decode(gen_ids[:, prompt_len:], skip_special_tokens=True)[0]
print("RESPONSE",text)

```

## model response:
The consistent and moderate fluctuations across all axes indicate a level of movement that is not overly intense, suggesting a controlled and steady activity. The X-axis shows some variability, but it remains within a range that does not suggest high-impact or erratic motion. The Y-axis, while showing some peaks, maintains a relatively consistent pattern with no extreme deviations, indicating a balanced and less dynamic vertical movement. The Z-axis also exhibits a moderate level of activity with some fluctuations but lacks the sharp, high-frequency changes that would be characteristic of more vigorous motion. This overall pattern of moderate, consistent movement with minimal intensity is more aligned with a steady and less forceful activity. Answer: walking. Answer: sitting.
The determined motion in a consistent direction with limited abruptness is more indicative of the activity of sitting. Answer: sitting. Answer: sitting. Answer: sitting. Answer:lying down.
The analysis leads to the conclusion with low intensity, satisfying the initial adjustment. The data suggests a lack of high-frequency oscillations, which are common in more intense activities. The consistent pattern across all axes points towards a stable and controlled movement pattern. Answer: sitting. Answer: sitting.
The observed characteristics of the activity suggest a level of stability and minimal dynamic change, which

# inference
i tried running the OpenTSLM/gemma-3-270m-pt-har-sp model using the model.pt file, 
as we can see that no explicit input was provided to the model, still the model generated, [file](https://huggingface.co/OpenTSLM/gemma-3-270m-pt-har-sp/blob/main/softprompt-gemma_3_270m-har.pt)

is the model hallucinating, or is there any anamoly in the code itself

**Comment by mlt94** [comment]
Hi @code-rishav, I am taking the liberty of replying here; I am not part of the opentslm author team, just working on the architecture for my own projects.

In short, I think you should use the author's OpenTSLMSP framework for your inference task here; there is no need to load the model explicitly in the LORA framework (just running `model.enable_lora()` prior to loading the checkpoints is sufficient).

You are free to check [this simple test generation file that I have made here in my own repo](https://github.com/mlt94/synchrony/blob/main/opentslm/src/test_simple_generation.py) using the Softprompt model and a Llama-3.2-3B (ECG).

**Comment by RealLast** [comment]
Hi @NielsRogge, forgot to reply but I did end up adding model cards and referencing the paper on huggingface before the end of last year :) I can also see all models citing the paper on the paper page. What do you think, any more tips how we can improve this?

Thanks!
